### ISSUE TITLE
Provide a list with unused mappings

### ISSUE THREAD

**Issue by motoras** [wiremock/wiremock#615]
As a project evolves, it usually ends up with a pretty large set of mappings files, some of them which are not in use any more, but it is very difficult to identify what files should be removed.

It will be great that either at shutdown or via a method call, wiremock to provide a list with the mappings it doesn't use anymore, either because, there is a newer file for the same url mapping, or because that mapping was no required in its current session.

**Comment by keithdtyler** [comment]
Your mappings could contain header and body matching as well, so unique url is not enough. It would have to be thorough.

Perhaps a status page showing statistics of mappings used would be easier and useful for cleaning up stale mappings.

**Comment by MasonM** [comment]
I wrote an extension for this: https://github.com/MasonM/wiremock-unused-stubs-extension

It adds 2 new API endpoints:
* `GET /__admin/unused_mappings`: retrieve any stub mapping not used to match any request in the request journal.
* `DELETE /__admin/unused_mappings`: delete all such mappings

@tomakehurst I can enter a PR to have this integrated into Wiremock, if you'd like. Doesn't matter to me; I'm fine maintaining the extension.

**Comment by tomakehurst** [comment]
I'm happy for it to stay as an extension if you are.

**Comment by romurta** [comment]
@MasonM Thx for the Extension :D

will 
DELETE /__admin/unused_mappings: delete all such mappings

Should delete the files from disk?
I am not seeing that happening. any advice?

**Comment by MasonM** [comment]
@romurta It just tried it and it deleted all persistent stub mappings from disk. Here's the steps I followed:
1. Run an instance of the standalone server: 
`java -jar build/libs/wiremock-unused-stubs-extension-0.2-standalone.jar`
2. Create a persistent mapping: 
`curl -d '{ "persistent": true, "request": { "method": "GET", "url": "/foo" }, "response": { "body": "FOO" } }' http://localhost:8080/__admin/mappings`
3. Delete unused mappings: 
`curl -s -X DELETE http://localhost:8080/__admin/unused_mappings`
4. List mappings (should be empty now): 
`ls mappings`

If that doesn't work for you, please enter an issue at https://github.com/MasonM/wiremock-unused-stubs-extension/issues

**Comment by jmayday** [comment]
Interesting thread. I haven't tried the extension yet, because I would rather like to have _unused_ endpoint available on Wiremock server started through Java. Possible use case scenario I see would be to add extra testcase, which would query the endpoint and fail the test if there are any unused mappings. Something like 
```
  @Order(Integer.MAX_VALUE)
  @Test
  @DisplayName("Fail if there are any unused Wiremock mappings")
  void shouldNotDetectAnyUnusedMappings() {

    String response = webClient //client using wiremock server
        .get()
        .uri(addr + "/__admin/requests/unmatched") //addr is the address of the server
        .retrieve()
        .bodyToMono(String.class) //map to something more convenient than String
        .block();

    //assert if there are any unused mappings and fail if necessary

  }
```

**Comment by Tom-Hudson** [comment]
Would it be possible to revisit this request? It looks like the extension isn't getting any maintenance?

**Comment by Tom-Hudson** [comment]
> @Tom-Hudson Not sure what maintenance you're referring to, but I consider the extension to be feature-complete, which is why I haven't made any changes.

@MasonM  - Maybe I made an unfair assumption that it wouldn't support WireMock v3?

I do still think this should be baked into the WireMock API though. It's a fairly common use case to check  stubs that have been created in a test are used.

**Comment by MasonM** [comment]
@Tom-Hudson You're right, it's probably not compatible. Enter an issue at https://github.com/MasonM/wiremock-unused-stubs-extension/issues and I'll update it.

**Comment by tomakehurst** [comment]
@MasonM how would you feel about bringing your extension into the core?

Perhaps one of you could create a PR?

**Comment by MasonM** [comment]
@tomakehurst Sure, done: https://github.com/wiremock/wiremock/pull/2991

**Comment by jmayday** [comment]
@tomakehurst hey, I'm having an issue trying to fetch unused mappings.

Isn't the new endpoint `/__admin/mappings/unmatched` conflicting with already existing `/__admin/mappings/{stubMappingId}`? I get a 400 bad request saying "unmatched is not a valid UUID" when trying to fetch unmatched mappings...
